packages/grid_client/src/manageURLs.ts

```diff
@@ -0,0 +1,123 @@
+    case "graphqlURL":
+      return { URLs, service: new GraphQLMonitor() };
+    case "activationURL":
+      return { URLs, service: new ActivationMonitor() };
+    default:
+      throw new GridClientError(`Can't resolve service with name "${name}"`);
+  }
+}
+
+/**
+ * Maps the result object to a new object based on the provided service names.
+ *
+ * This function takes a result object and an array of service names. It returns
+ * a new object where the keys are the provided service names and the values are taken from
+ * the result object, using a mapping from `MapServiceNames`.
+ *
+ * @param {Object.<string, string | null>} result - The result object to map from.
+ * @param {string[]} services - An array of service names to map.
+ * @returns {Object.<string, string | null>} A new object with the mapped service names and values.
+ */
+function mapResult(result: { [key: string]: string | null }, services: string[]) {
+  return services.reduce((acc, service) => {
+    acc[service] = result[MapServiceNames[service]];
+    return acc;
+  }, {});
+}
+
+/**
+ * Retrieves and assigns available URLs for services missing URLs in the client options.
+ *
+ * This function checks which service URLs are missing in the provided client options,
+ * prepares the services with predefined URLs, and uses a `ServiceUrlManager` to
+ * retrieve available service URLs. The retrieved URLs are then assigned back to the
+ * client options.
+ *
+ * @param {ClientOptions} clientOptions - The client options containing service URLs and network environment.
+ * @returns {Promise<void>} A promise that resolves when the URLs have been retrieved and assigned.
+ */
+export async function getAvailableURLs(clientOptions: ClientOptions) {
+  const { proxyURL, relayURL, substrateURL, graphqlURL, activationURL, network } = clientOptions;
+  const currentURLs = { proxyURL, relayURL, substrateURL, graphqlURL, activationURL };
+  const base = network === NetworkEnv.main ? "grid.tf" : `${network}.grid.tf`;
+  const URLS = {
+    relayURL: [`wss://relay.${base}`, `wss://relay.02.${base}`],
+    proxyURL: [`https://gridproxy.${base}`, `https://gridproxy.02.${base}`],
+    activationURL: [
+      `https://activation.${base}/activation/activate`,
+      `https://activation.02.${base}/activation/activate`,
+    ],
+    graphqlURL: [`https://graphql.${base}/graphql`, `https://graphql.02.${base}/graphql`],
+    substrateURL: [`wss://tfchain.${base}/ws`, `wss://tfchain.02.${base}/ws`],
+  };
+  const missingServicesURLS = getServicesWithoutURLs(currentURLs);
+
+  if (missingServicesURLS.length == 0) return;
+
+  const services = missingServicesURLS.reduce((acc, servicesName) => {
+    acc.push(prepareServices(servicesName, URLS[servicesName]));
+    return acc;
+  }, [] as Service[]);
+  const result = await new ServiceUrlManager({
+    services: services,
+    silent: true,
+  }).getAvailableServicesStack();
+  Object.assign(clientOptions, mapResult(result, missingServicesURLS));
+}
```

Comment:
I think mutating object byRef might lead to unexpected behaviour specially in vue apps as it turns objects into proxy which rerender ui whenever it gets mutated

So I think we should avoid update objects byref